src/Pos/Wallet/Web/Tracking.hs

```diff
@@ -248,67 +290,63 @@ syncWalletWithGStateUnsafe encSK = do
                      -- or if the application was interrupted during rollback.
                      -- We don't load blocks explicitly, because blockain can be long.
                      maybe (pure mempty)
-                         (\wNextHeader -> foldlUpWhileM (applyBlock allAddresses) wNextHeader constTrue mappendR mempty)
+                         (\wNextH ->
+                            foldlUpWhileM (applyBlock allAddresses) wNextH loadCond mappendR mempty)
                          =<< resolveForwardLink wHeader
-               | diff tipHeader < diff wHeader -> do
+               | diff gstateH < diff wHeader -> do
                      -- This rollback can occur
                      -- if the application was interrupted during blocks application.
                      blunds <- getNewestFirst <$>
-                         DB.loadBlundsWhile (\b -> getBlockHeader b /= tipHeader) (headerHash wHeader)
+                         DB.loadBlundsWhile (\b -> getBlockHeader b /= gstateH) (headerHash wHeader)
                      pure $ foldl' (\r b -> r <> rollbackBlock allAddresses b) mempty blunds
-               | otherwise -> mempty <$ logInfo (sformat ("Wallet "%build%" is already synced") wAddr)
-
-    whenJustM (WS.getWalletSyncTip wAddr) $ \wTip ->
-        if | wTip == genesisHash && headerHash tipHeader == genesisHash ->
-               logDebug $ sformat ("Wallet "%build%" at genesis state, synced") wAddr
-           | otherwise -> sync wTip
+               | otherwise -> mempty <$ logDebug (sformat ("Wallet "%build%" is already synced") wAddr)
+
+    startFromH <- maybe firstGenesisHeader pure wTipHeader
+    genesisUtxo <- genesisUtxoM
+    mapModifier@CAccModifier{..} <- computeAccModifier startFromH
+    whenNothing_ wTipHeader $ do
+        let encInfo = getEncInfo encSK
+        let ownGenesisUtxo =
+                M.fromList $
+                map fst $
+                selectOwnAccounts encInfo (txOutAddress . toaOut . snd) (M.toList genesisUtxo)
+        WS.getWalletUtxo >>= WS.setWalletUtxo . (ownGenesisUtxo <>)
+    applyModifierToWallet wAddr gstateHHash mapModifier
+    logInfo $ sformat ("Wallet "%build%" has been synced with tip "
+                        %shortHashF%", "%build)
+        wAddr (maybe genesisHash headerHash wTipHeader) mapModifier
+    where
+      firstGenesisHeader :: m (BlockHeader ssc)
+      firstGenesisHeader = resolveForwardLink (genesisHash @BlockHeaderStub) >>=
+          maybe (error "Unexpected state: genesisHash doesn't have forward link")
+              (maybe (error "No genesis block corresponding to header hash") pure <=< DB.blkGetHeader)
+
+
+runWithWalletUtxo
+    :: (MonadReader ctx m, HasLens GenesisUtxo ctx GenesisUtxo, WebWalletModeDB ctx m)
+    => ToilT () (GenesisToil m) a
+    -> m a
+runWithWalletUtxo action = do
+    walletUtxo <- WS.getWalletUtxo
+    runGenesisToil $ fst <$> runToilTLocal (fromUtxo walletUtxo) def mempty action
```

Comment:
So, instead of using `ToilT` over `DBToil` while getting `CAccModifier`, you use it over `GenesisToil`? This will break transaction history. If there's an incoming transactions in block which inputs don't belong to our wallet and aren't from genesis utxo, these inputs won't be found. So you cannot get rid of `DBToil` here.
You can still use the idea of pre-fetching locally cached utxo to reduce db queries, though.